bin/solve_stats.py

```diff
@@ -0,0 +1,141 @@
+from os import makedirs
+
+import matplotlib.pyplot as plt
+
+from contest import call_api, get_contest_id
+from util import ProgressBar
+
+
+def generate_solve_stats(post_freeze):
+    contest_id = get_contest_id()
+
+    # The endpoint should not start with a slash
+    def req(endpoint):
+        url = f'/contests/{contest_id}/{endpoint}'
+        bar.start(url)
+        r = call_api('GET', url)
+        r.raise_for_status()
+        bar.done()
+        try:
+            return r.json()
+        except Exception as e:
+            print(f'\nError in decoding JSON:\n{e}\n{r.text()}')
+
+    # Turns an endpoint list result into an object, mapped by 'id'
+    def req_assoc(endpoint):
+        return {o['id']: o for o in req(endpoint)}
+
+    def time_string_to_minutes(time_string):
+        hours, minutes, seconds = (time_string or '0:0:0').split(':')
+        return int(hours) * 60 + int(minutes) + float(seconds) / 60
+
+    bar = ProgressBar('Fetching', count=7, max_len=28 + len(contest_id))
+
+    contest = req('')
+    freeze_duration = time_string_to_minutes(contest['scoreboard_freeze_duration'])
+    contest_duration = time_string_to_minutes(contest['duration'])
+    bins = 120
+    scale = contest_duration / bins
+
+    problems = req_assoc('problems')
+    submissions = req_assoc('submissions')
+    teams = req_assoc(f'teams?public=1')
+    languages = req_assoc('languages')
+    judgement_types = req_assoc('judgement-types')
+    judgement_types[''] = {'id': '', 'name': 'pending'}
+    judgement_colors = {'AC': 'lime', 'WA': 'red', 'TLE': '#c0f', 'RTE': 'orange', '': 'skyblue'}
+
+    for j in req('judgements'):
+        # Note that the submissions list only contains submissions that were submitted on time,
+        # while the judgements list contains all judgements, therefore the submission might not exist.
+        if j['valid'] and j['submission_id'] in submissions:
+            # Merge judgement with submission. Keys of judgement are overwritten by keys of submission.
+            submissions[j['submission_id']] = {**j, **submissions[j['submission_id']]}
```

Comment:
what is happening here? Can you not do something like `submissions[id]['judgement'] = j`?

How does this handle multiple judgements for a single submission? Is `'valid'` the one that is 'active'?

Reply:
That's probably cleaner indeed, thanks for the suggestion! :slightly_smiling_face: 

The DOMjudge API does not state anything about the `'valid'` field, but from what I've seen, I think your idea is correct.

Reply:
ok can you add the thing about 'valid' as a comment

Reply:
Sure, done :slightly_smiling_face: